### PR TITLE
 Fix bin boundaries calculation in Discretization layer

### PR DESCRIPTION
**Pull request by ul611**
### Problem
This PR fixes an issue where `Discretization` layer produced uneven bin boundaries, resulting in the last bin containing significantly more elements than others. After investigation, the root cause was found in `compress_summary` function: `cum_weight_percents` was normalized by the second-to-last value (`cum_weights[-1]`), not the last, as the last value is excluded in `summarize` function (variable `boundaries`). 

### Solution
The issue is fixed by including the last value in the `summarize` function. This is achieved by shifting the boundaries one position to the left: `boundaries = values[int(start) - 1 :: int(step)]`. To ensure that `int(start) - 1` is never negative, start is set to `start = max(increment, 1)`.

### Visualization

```Python
import tensorflow  as tf
import matplotlib.pyplot as plt

n = 10000
viz_step = 1  # increase to speed up

# array
x = tf.random.uniform((n,))

for num_bins in range(1, 101, viz_step):

    discretization = tf.keras.layers.Discretization(num_bins=num_bins)
    discretization.adapt(x)
    y = discretization(x)

    fig, axis = plt.subplots(1, 1, figsize=(4, 4))
    counts,bins, _ = axis.hist(y, bins=num_bins, rwidth=0.5)
    plt.title(num_bins)
    plt.show()

# batched dataset
batch1 = np.random.uniform(0, 10, n).astype(np.float32)
batch2 = np.random.uniform(5, 15, n).astype(np.float32)
batch3 = np.random.uniform(150, 250, n).astype(np.float32)

x = tf.data.Dataset.from_tensor_slices([batch1, batch2, batch3])
all_data = np.concatenate([batch1, batch2, batch3])

for num_bins in range(1, 101, viz_step):
    discretization = tf.keras.layers.Discretization(num_bins=num_bins)
    discretization.adapt(x)
    y = discretization(all_data).numpy()

    fig, axis = plt.subplots(1, 1, figsize=(4, 4))
    counts,bins, _ = axis.hist(y, bins=num_bins, rwidth=0.5)
    plt.title(num_bins)
    plt.show()
```

## Contributor Agreement

**Please review our [AI-Assisted Contribution Policy](../CONTRIBUTING.md#ai-assisted-contribution-policy) and check all boxes below before submitting your PR for review:**

- [x] I am a human, and not a bot.
- [x] I will be responsible for responding to review comments in a timely manner.
- [x] I will work with the maintainers to push this PR forward until submission.

**Note:** Failing to adhere to this agreement may result in your future PRs no longer being reviewed.

### CODE REVIEW

**Comment by ul611** [comment]
@gemini-code-assist review

**Comment by ul611** [comment]
@gemini-code-assist review

File: keras/src/layers/preprocessing/discretization.py [hunk 1]
```diff
@@ -333,9 +333,9 @@ def summarize(values, epsilon):
     elements = np.size(values)
     num_buckets = 1.0 / epsilon
     increment = elements / num_buckets
-    start = increment
-    step = max(increment, 1)
-    boundaries = values[int(start) :: int(step)]
+    start = max(increment, 1)
+    step = start
+    boundaries = values[int(start) - 1 :: int(step)]
```
Comment (hertschuh) [inline keras/src/layers/preprocessing/discretization.py]: Just to make it a little less confusing, can you rewrite it like this:

```python
    step = max(increment, 1)  
    start = step - 1
    boundaries = values[int(start) :: int(step)]
```

**Review by hertschuh — CHANGES_REQUESTED** [review]
I _think_ the math is correct. But can you re-run your test but with a non-uniform distribution, like `np.random.normal`?

File: keras/src/layers/preprocessing/discretization.py (hunk 1 above)
Comment (ul611) [reply keras/src/layers/preprocessing/discretization.py]: Done

**Comment by ul611** [comment]
Thanks for the suggestion. I tested on normal distribution: the original layer produces tail boundary shifts up to `-0.25 ` (or even lower for a single batch) compared to theoretical, while the fixed version stays within `±0.03`. Code snippet and screenshots attached. The difference is negligible for small `num_bins` but significant for `num_bins ~ 1/epsilon`.

image

image

```Python
import numpy as np
import tensorflow as tf
import matplotlib.pyplot as plt
from scipy.stats import norm

epsilon = 0.01
n = 40000
np.random.seed(42)

for num_bins in range(5, 100, 5):
    # Data (normal distribution)
    batch1 = np.random.normal(0, 1, n).astype(np.float32)
    batch2 = np.random.normal(0, 1, n).astype(np.float32)
    batch3 = np.random.normal(0, 1, n).astype(np.float32)
    all_data = np.concatenate([batch1, batch2, batch3])
    x_ds = tf.data.Dataset.from_tensor_slices([batch1, batch2, batch3])

    # Normal distribution parameters calculated from all data
    mu = np.mean(all_data)
    std = np.std(all_data, ddof=1)

    # Theoretical quantiles (ideal boundaries)
    quantiles = np.linspace(0, 1, num_bins + 1)[1:-1]
    theoretical = norm.ppf(quantiles, loc=mu, scale=std)

    # Layers
    original_layer = tf.keras.layers.Discretization(num_bins=num_bins, epsilon=epsilon)
    original_layer.adapt(x_ds)
    orig_bounds = np.array(original_layer.bin_boundaries)
    dev_orig = orig_bounds - theoretical

    # Plot deviations
    plt.figure(figsize=(12, 5))
    x = np.arange(1, len(theoretical)+1)
    plt.plot(x, dev_orig, 'o-', label='Original', markersize=3, alpha=0.7)
    plt.axhline(0, color='k', linewidth=0.8)
    plt.xlabel('Boundary index')
    plt.ylabel('Deviation from theoretical quantile')
    plt.title(f'Normal distribution, num_bins={num_bins}')
    plt.legend()
    plt.grid(True)
    plt.show()
    ```

**Review by hertschuh — APPROVED** [review]
Thanks for investigating this!
